src/Sarif/Visitors/InsertOptionalDataVisitor.cs

```diff
@@ -62,6 +68,28 @@ public override Run VisitRun(Run node)
             return visited;
         }
 
+        private void InsertVersionControlInformation()
+        {
+            var gitInformation = new GitInformation(s_fileSystem);
+            string currentDirectory = s_executionEnvironment.CurrentDirectory;
```

Comment:
>CurrentDirectory [](start = 61, length = 16)

This will work some of the time but seems haphazard. What we need to happen is for every physical location that is included in the log file be examined for relevance to source control.

Reply:
`ENVIRONMENT_DEPENDENT_ENLISTMENT_ROOT`?

---
In reply to: [475661249](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475661249) [](ancestors = 475661249)

Reply:
This is the directory from which `GitInformation.GetRepositoryRoot` starts its upward search for the directory containing `.git`. The assumption was that this operation works only if you're running from somewhere under the repo root.

**BUT**: Now I see in `VisitPhysicalLocation` below, where it is enriching the regions and filling in the snippets, that it relies on `originalUriBaseIds` to resolve URIs -- and it if can't resolve a URI it silently fails to fill in the information.

So how about this: We try the operation starting with each entry in `originalUriBaseIds`. If all of them fail (that is, if there is no entry that points at or into a directory containing a `.git`), only then do we throw. (And we _do_ throw; I don't think it's ok to silently do nothing here.) We have to be careful in the presence of submodules to pick the topmost directory with a `.git`.

---
In reply to: [475661249](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475661249) [](ancestors = 475661249)

Reply:
(Having said that, I don't understand what you mean by examining every physical location. Do you mean "See if any physical location mentioned anywhere in the log file is an absolute URI from which the repo root can be determined?")

---
In reply to: [475691019](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475691019) [](ancestors = 475691019,475661249)

Reply:
Yes. The goal of this feature is that any physical location referenced in the log file is expressed as a relative path with corresponding enlistment root, if that PLC is in fact a source-controlled item. The code as authored is not accomplishing this. 

---
In reply to: [475692378](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475692378) [](ancestors = 475692378,475691019,475661249)

Reply:
There are two distinct issues here:

1. How do we determine the repo root (and thus populate `run.versionControlProvenance`)? My implementation relies on the current directory; my alternative proposal above would rely on `originalUriBaseIds`, similar to what `VisitPhysicalLocation` does below.

2. Having determined the repo root, how do we express all URIs with respect to it? We're 100% agreed that we need to do that. **BUT**: as the Multitool is currently structured, that's not the job of the `rewrite` command/`InsertOptionalDataVisitor`. Rather, it's the job of the `rebaseuri` command/`RebaseUriVisitor`.

So IMO the code is structured correctly, and what the normalizer should do is invoke both the `rewrite` command and the `rebaseuri` command -- **EXCEPT** it turns out that the `rebaseuri` command supports the options for, and invokes, the `InsertOptionalDataVisitor`, and so it's actually a superset of the `rewrite` command, so a single invocation of `rebaseuri` will suffice.

This gets into the whole multitool redesign discussion. We can choose to do it now, or after the normalizer is done. But either way, the URI rebasing doesn't belong in the `InsertOptionalDataVisitor`.

---
In reply to: [475696046](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475696046) [](ancestors = 475696046,475692378,475691019,475661249)

Reply:
I suggest we get this code in now. As we discussed offline, commands like rebaseuri are redudant and more granular than we need. 

I will put some thought into the larger design.

Reply:
Ok, I will merge what I have.

---
In reply to: [475714891](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475714891) [](ancestors = 475714891)

Reply:
I mean, "Ok, please merge what I have."

---
In reply to: [475769205](https://github.com/microsoft/sarif-sdk/pull/2035#discussion_r475769205) [](ancestors = 475769205,475714891)